snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,271 @@
+from snuba.utils.streams.consumers.types import Message, TOffset, TStream, TValue
+
+
+ConsumerGroup = str
+
+
+@dataclass
+class CommitData(Generic[TStream, TOffset]):
+    consumer_group: ConsumerGroup
+    stream: TStream
+    offset: TOffset
+
+
+# TODO: This should be moved into the abstract consumer backend.
+class Topic(ABC, Generic[TStream]):
+    @abstractmethod
+    def __contains__(self, stream: TStream) -> bool:
+        raise NotImplementedError
+
+
+@dataclass
+class Subscription(Generic[TStream, TOffset]):
+    topics: Sequence[Topic[TStream]]
+    remote_consumer_group_offsets: MutableMapping[
+        TStream, MutableMapping[ConsumerGroup, TOffset]
+    ] = field(default_factory=lambda: defaultdict(dict))
+
+
+# TODO: TOffset needs to be bound as some sort of comparable type.
+
+
+class SynchronizedConsumerBackend(ConsumerBackend[TStream, TOffset, TValue]):
+    def __init__(
+        self,
+        backend: ConsumerBackend[TStream, TOffset, TValue],
+        commit_log_consumer_backend: ConsumerBackend[
+            # NOTE: These TStream and TOffset values for the commit log
+            # consumer do not need to match, but this simplifies things a bit
+            # so that we don't have to declare two additional type variables.
+            TStream,
+            TOffset,
+            CommitData[TStream, TOffset],
+        ],
+        remote_consumer_groups: Set[ConsumerGroup],
+        get_commit_log_streams_for_topic: Callable[[Topic[TStream]], Sequence[TStream]],
+    ) -> None:
+        self.__backend = backend
+        self.__commit_log_consumer_backend = commit_log_consumer_backend
+        self.__remote_consumer_groups = remote_consumer_groups
+        self.__get_commit_log_streams_for_topic = get_commit_log_streams_for_topic
+
+        self.__lock = Lock()
+        self.__subscription: Subscription[TStream, TOffset] = Subscription([])
+        self.__shutdown_requested = Event()
+
+        # TODO: If the commit log consumer crashes, it should cause all method
+        # calls on this object to raise (probably a RuntimeError raised from
+        # the commit log consumer exception.)
+        self.__commit_log_consumer_future = execute(self.__run_commit_log_consumer)
+
+    def __run_commit_log_consumer(self) -> None:
+        # TODO: How does the commit log consumer ensure that it is the only
+        # consumer within it's consumer group (or that it otherwise has the
+        # comprehensive set of partitions assigned for the commit log topic?)
+        # TODO: How do we ensure that the commit log consumer is not committing
+        # it's offsets (or is subject to automatic offset resets?)
+
+        # XXX: This reference to the subscription object is only maintained so
+        # that it can be referenced checked during the commit log consumer main
+        # loop. It should not be used otherwise!
+        current_subscription: Optional[Subscription[TStream, TOffset]] = None
```

Comment:
This is pretty dangerous, but I couldn't think of another good way to do it. Keeping the object reference alive (which this does) is the only way to prevent the `id` from being recycled, which is the only way to check if the assignment has changed, unless we just want to compare the equality of subscription topic sets (which could also work, I think.)